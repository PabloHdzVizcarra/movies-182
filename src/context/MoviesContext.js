import React, { createContext, useReducer, useCallback, useState } from 'react';
import { moviesReducer } from '../reducers/moviesReducer';
import { types } from '../types/types';
import { db, firebase } from '../libs/firebase';

export const MoviesContext = createContext();



export const MoviesProvider = ({ children }) => {

  const [state, dispatch] = useReducer(moviesReducer, {
    movies: [],
    favoriteMovies: [],
    searchMovie: [],
    movie: {},
    loadingMovies: true
  });

  const [errorForm, setErrorForm] = useState({
    error: false,
    message: ''
  });
  
  const { movies, favoriteMovies, searchMovie, movie, loadingMovies } = state;
  

  const getTopRatedMovies = useCallback(async () => {
    try {
      const apiKey = process.env.REACT_APP_APIKEY;
      
      console.log('OBTENIENDO PELICULAS');
      const data = await fetch(`https://api.themoviedb.org/3/movie/top_rated?api_key=${apiKey}&language=es-MX&page=1`);
      const { results } = await data.json();
      
      dispatch({
        type: types.loadingTopRatedMovies,
        payload: results
      });
    } catch (error) {
      console.log(error);
    }
  }, []);

  const getPopularMovies = useCallback(async () => {
    try {
      const apiKey = process.env.REACT_APP_APIKEY;
      
      console.log('OBTENIENDO PELICULAS');
      const data = await fetch(`https://api.themoviedb.org/3/movie/popular?api_key=${apiKey}&language=es-MX&page=1`);
      const { results } = await data.json();
      
      dispatch({
        type: types.loadingPopularMovies,
        payload: results
      });
    } catch (error) {
      console.log(error);
    }
  }, []);

  const getUpcomingMovies = useCallback(async () => {

    try {
      const apiKey = process.env.REACT_APP_APIKEY;
      
      console.log('OBTENIENDO PELICULAS');
      const data = await fetch(`https://api.themoviedb.org/3/movie/upcoming?api_key=${apiKey}&language=es-MX&page=1`);
      const { results } = await data.json();
      
      dispatch({
        type: types.loadingUpcomingMovies,
        payload: results
      });
    } catch (error) {
      console.log(error);
    }
  }, []);

  const getMoviesFirebase = async(uid) => {

    try {
      
      let allFavoritesMovies = [];

      await db.collection(uid)
      .doc("movies")
      .collection("favorites")
      .get()
      .then((querySnapshot) => {
        querySnapshot.forEach((doc) => {
          let movie;
          movie = { ...doc.data() }
          movie.docID = doc.id;
          allFavoritesMovies.push(movie)
        });
      });

      // const reversedArray = allFavoritesMovies.reverse();

      dispatch({
        type: types.getFavoritesMoviesFirebase,
        payload: allFavoritesMovies
      })

    } catch (error) {
      console.log(error);
    }
  }

  const getMovieSearch = async (key) => {

    try {
      const apiKey = process.env.REACT_APP_APIKEY;
      const data = await fetch(`https://api.themoviedb.org/3/search/movie?api_key=${apiKey}&query=${key}`);
      
      const { results } = await data.json();
      console.log(results);

      if (results.length === 0) {
        return setErrorForm({
          error: true,
          message: 'Intenta buscando otra pelicula'
        })
      }

      setErrorForm({
        error: false,
        message: ''
      })

      dispatch({
        type: types.searchMovies,
        payload: results
      });

    } catch (error) {
      console.log(error);
    }
  }

  const getMovieID = async (movieID) => {

    try {

      const apiKey = process.env.REACT_APP_APIKEY;
      const data = await fetch(`https://api.themoviedb.org/3/movie/${movieID}?api_key=${apiKey}&language=es-ES`);

      const resp = await data.json();
      
      dispatch({
        type: types.selectedMovie,
        payload: resp
      })
      
    } catch (error) {
      console.log(error);
    }
  }



  const addFavoriteMovie = async (movie) => {

    const {
      poster_path,
      id,
      genres,
      original_title,
      original_language,
      overview,
      popularity,
      production_countries,
      release_date,
      tagline,
      title,
      vote_average,
      vote_count
    } = movie;

    const imgMovie = `https://image.tmdb.org/t/p/w500/${poster_path}`;

    const favoriteMovie = {
      imgMovie,
      id,
      genres,
      original_title,
      original_language,
      overview,
      popularity,
      production_countries,
      release_date,
      tagline,
      title,
      vote_average,
      vote_count
    }

    const { uid } = firebase.auth().currentUser;

    if (uid) {

      const { id } = await db.collection(uid).doc('movies').collection('favorites').add(favoriteMovie);

      favoriteMovie.firebaseID = id;

      dispatch({
        type: types.addFavoritesMovies,
        payload: favoriteMovie
      })
      
      console.log('PELICULA AGREGADA A FAVORITOS');

    } else {
      throw new Error('No hay usuario autenticado')
    }

  }

  const deleteMovieFromFavorites = async (userID, docID) => {
    console.log(docID);

    dispatch({
      type: types.removeMovieFromFavorites,
      payload: docID
    })
    

    await db.collection(userID).doc('movies').collection('favorites').doc(docID)
      .delete();

    return console.log(`PELICULA ELIMINADA`);
  }
  
  return (
    <MoviesContext.Provider value={{
      movie,
      movies,
      favoriteMovies,
      searchMovie,
      loadingMovies,
      errorForm,
      getMovieID,
      getPopularMovies,
      getTopRatedMovies,
      getUpcomingMovies,
      getMovieSearch,
      addFavoriteMovie,
      getMoviesFirebase,
      deleteMovieFromFavorites,
      dispatch
    }}>
      {children}
    </MoviesContext.Provider>
  )
}