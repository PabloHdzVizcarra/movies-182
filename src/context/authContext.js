import React from 'react'
import { types } from '../types/types';

const AuthStateContext = React.createContext();
const AuthDispatchContext = React.createContext();

function authReducer(state, action) {
  switch (action.type) {

    case types.setActiveUser:
      return {
        ...state,
        activeUser: action.payload,
        isActived: true
      }

    case types.logoutUser:
      return {
        ...state,
        activeUser: {},
        isActived: false
      }
    default: {
      throw new Error(`Unhandled action type: ${action.type}`)
    }
  }
}


function AuthProvider({ children }) {
  const [state, dispatch] = React.useReducer(authReducer, {
    activeUser: {},
    isActived: false
  });

  return (
    <AuthStateContext.Provider value={state}>
      <AuthDispatchContext.Provider value={dispatch}>
        {children}
      </AuthDispatchContext.Provider>
    </AuthStateContext.Provider>
  )
}

function useAuthState() {
  const context = React.useContext(AuthStateContext);

  if (context === undefined) {
    throw new Error('useAuthState must be used within AuthProvider')
  }

  return context
}

function useAuthDispatch() {
  const context = React.useContext(AuthDispatchContext);

  if (context === undefined) {
    throw new Error('useAuthDispatch must be used within AuthProvider')
  }

  return context
}

export {
  AuthProvider,
  useAuthState,
  useAuthDispatch
}