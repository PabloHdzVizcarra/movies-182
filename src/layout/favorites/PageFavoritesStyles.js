import styled from 'styled-components';
import { respondTo } from '../../styles/_respondTo';

export const MoviesContainer = styled.div`
  display: grid;
  gap: 10px;
  padding: 16px;
  height: min-content;
  background-color: #FFFFFF;

  ${respondTo.xs`
    grid-template-columns: 1fr;
  `}

  ${respondTo.sm`
    grid-template-columns: 1fr 1fr;
  `}

  ${respondTo.md`
    grid-template-columns: 1fr 1fr 1fr 1fr;
  `}

  ${respondTo.lg`
    grid-template-columns: 1fr 1fr 1fr 1fr 1fr 1fr;
  `}
`;