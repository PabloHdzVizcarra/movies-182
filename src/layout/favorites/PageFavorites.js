import React, { useContext, useEffect } from 'react'
import { useAuthState } from '../../context/authContext'
import { MoviesContext } from '../../context/MoviesContext';
import { MovieCard } from '../../components/movies/MovieCard';
import styled from 'styled-components';
import { respondTo } from '../../styles/_respondTo';

const MoviesContainer = styled.div`
  display: grid;
  gap: 10px;
  padding: 16px;
  height: max-content;
  background-color: rgba(133, 133, 133, .2);
  min-height: 87vh;

  ${respondTo.xs`
    grid-template-columns: 1fr;
  `}

  ${respondTo.sm`
    grid-template-columns: 1fr 1fr;
  `}

  ${respondTo.md`
    grid-template-columns: 1fr 1fr 1fr 1fr;
  `}

  ${respondTo.lg`
    grid-template-columns: 1fr 1fr 1fr 1fr 1fr 1fr;
  `}
`;

export const PageFavorites = () => {

  const data = useAuthState();
  const { getMoviesFirebase, favoriteMovies, loadingMovies } = useContext(MoviesContext);

  useEffect(() => {
    if (data.activeUser.uid) {
      getMoviesFirebase(data.activeUser.uid);

    }
    //eslint-disable-next-line
  }, [data.activeUser.uid]);
  
  if (loadingMovies) {

    return (
      <p>Loading Page</p>
    )
  }
  return (
    <MoviesContainer>
      {(favoriteMovies.length === 0)
        ? <p>No tienes peliculas favoritas</p>
        : favoriteMovies.map((movie) => (
          <MovieCard
            isActived={data.isActived}
            key={movie.id}
            {...movie}
          />
      ))}
    </MoviesContainer>
  )
}
