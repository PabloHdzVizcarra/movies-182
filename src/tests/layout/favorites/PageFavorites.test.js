import React from 'react';
import "@testing-library/jest-dom";
import { mount } from 'enzyme';
import { PageFavorites } from '../../../layout/favorites/PageFavorites';
import { AuthProvider } from '../../../context/authContext';
import { MoviesContext } from '../../../context/MoviesContext';
import { MemoryRouter } from 'react-router-dom';

describe('Test in <PageFavorites />', () => {

  test('if there are no favorite movies the component must render the text "No tienes peliculas favoritas"', () => {
    
    const wrapper = mount(
      <AuthProvider>
        <MoviesContext.Provider value={{
          loadingMovies: false,
          favoriteMovies: []
        }}>          
          <PageFavorites />
        </MoviesContext.Provider>
      </AuthProvider>
    )

    const text = wrapper.find('p').text().trim();
    expect(text).toBe('No tienes peliculas favoritas');
  })

  test('if the component has favorite movies these should be rendered', () => {

    const mockMoviesContext = {
      loadingMovies: false,
      favoriteMovies: [
        {
          title: "Harry Potter",
          id: 612706
        }
      ]
    }
    const wrapper = mount(
      <AuthProvider>
        <MoviesContext.Provider value={mockMoviesContext}>
          <MemoryRouter>
            <PageFavorites />
          </MemoryRouter>  
        </MoviesContext.Provider>
      </AuthProvider>
    )

    
    expect(wrapper).toMatchSnapshot();
    const node = wrapper.find('h4').text().trim();
    expect(node).toBe(mockMoviesContext.favoriteMovies[0].title);
  })
  
  
})
