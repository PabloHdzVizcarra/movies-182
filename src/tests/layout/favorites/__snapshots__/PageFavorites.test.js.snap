// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <PageFavorites /> if the component has favorite movies these should be rendered 1`] = `
<div
  className="sc-AxgMl MySFY"
>
  <div
    className="sc-AxjAm kuIkN animate__animated animate__fadeIn"
  >
    <div>
      <a
        href="/search/612706"
        onClick={[Function]}
      >
        <img
          alt="No hay Imagen Disponible"
          src="https://images.unsplash.com/photo-1518676590629-3dcbd9c5a5c9?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=634&q=80"
        />
      </a>
    </div>
    <div>
      <div
        className="sc-AxhCb buhHwi"
      >
        <h4>
          Harry Potter
        </h4>
        <div
          className="sc-AxirZ dZnDSZ"
        >
          <div>
            <p>
              Calificacion:
            </p>
            <p>
              <b />
               
            </p>
          </div>
        </div>
        <div
          className="sc-AxirZ dZnDSZ"
        >
          <div>
            <p>
               Likes: 
            </p>
            <p>
              <b />
               
            </p>
          </div>
        </div>
        <div
          className="sc-AxirZ dZnDSZ"
        >
          <div
            style={
              Object {
                "paddingBottom": "10px",
              }
            }
          >
            <p>
              Estreno: 
            </p>
            <p>
              <b>
                12 agosto 2020
              </b>
               
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
